instrumentation/undertow-1.4/src/main/java/io/opentelemetry/javaagent/instrumentation/hypertrace/undertow/v1_4/utils/Utils.java

```diff
@@ -0,0 +1,63 @@
+/*
+ * Copyright The Hypertrace Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.opentelemetry.javaagent.instrumentation.hypertrace.undertow.v1_4.utils;
+
+import io.undertow.server.HttpServerExchange;
+import java.io.UnsupportedEncodingException;
+import java.nio.ByteBuffer;
+import java.nio.charset.Charset;
+import org.hypertrace.agent.core.instrumentation.HypertraceSemanticAttributes;
+import org.hypertrace.agent.core.instrumentation.SpanAndObjectPair;
+import org.hypertrace.agent.core.instrumentation.buffer.BoundedBuffersFactory;
+import org.hypertrace.agent.core.instrumentation.buffer.BoundedByteArrayOutputStream;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public final class Utils {
+
+  public static void handleTransfer(
+      final ByteBuffer byteBuffer, int numBytesRead, SpanAndObjectPair spanAndObjectPair) {
+    final Object associatedObject = spanAndObjectPair.getAssociatedObject();
+    final HttpServerExchange exchange;
+    if (associatedObject instanceof HttpServerExchange) {
+      exchange = (HttpServerExchange) associatedObject;
+    } else {
+      return;
+    }
+
+    if (!exchange.isRequestComplete()) {
+      return;
+    }
+    final BoundedByteArrayOutputStream boundedByteArrayOutputStream =
+        BoundedBuffersFactory.createStream(
+            byteBuffer.remaining(), Charset.forName(exchange.getRequestCharset()));
+    final byte[] dst = new byte[byteBuffer.remaining()];
+    byteBuffer.get(dst);
+    boundedByteArrayOutputStream.write(dst, 0, dst.length);
+    log.info("test");
+    final String body;
+    try {
+      body = boundedByteArrayOutputStream.toStringWithSuppliedCharset().trim();
+    } catch (UnsupportedEncodingException e) {
+      throw new IllegalArgumentException("illegal encoding", e);
```

Comment:
can this happen? wasn't the correct encoding parsed before?

Reply:
Yeah, you're correct. This exception is propagated by `toStringWithSuppliedCharset`. I think I should modify `toStringWithSuppliedCharset` to catch the `UnsupportedEncodingException` and wrap it in an unchecked exception, since `toStringWithSuppliedCharset` calls `toString(String)` with the charset name retrieved from a `Charset` instance.  So
```
  public synchronized String toStringWithSuppliedCharset() throws UnsupportedEncodingException {
    return super.toString(this.charset.name());
  }
```
  
 Would  become: 
```
public synchronized String toStringWithSuppliedCharset() {
    try {
      return super.toString(this.charset.name());
    } catch (UnsupportedEncodingException e) {
      throw new IllegalStateException("Charset name is guaranteed to be valid");
    }
  }
```

Reply:
I would rather avoid throwing the exception and instead just add a log. The agent should not break the app in any case. 

On a different note I don't recall why the method is synchronized.

Reply:
Works for me! 👍 Yeah, I'm not sure why it is either, I'll look into it in the future but for now I'm going to hold off on changing it since some of our users are waiting on the instrumentation fix for undertow 😃